### PR TITLE
add the root type expand impl for rollup syntax.

### PR DESCRIPTION
**Pull request by AilinKid**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v8.3 (TiDB 8.3 versions)
- [ ] v8.2 (TiDB 8.2 versions)
- [ ] v8.1 (TiDB 8.1 versions)
- [ ] v8.0 (TiDB 8.0 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: functions-and-operators/group-by-modifier.md [hunk 1]
```diff
@@ -34,7 +34,7 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
 
 ## 准备条件
 
-目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
+目前，TiDB 支持在 TiDB 自身和 TiFlash MPP 模式下为 `WITH ROLLUP` 语法生成有效的执行计划，可以在执行计划中通过观察 `EXPAND` 算子存储引擎属性来辨别。`EXPAND` 的 MPP 属性要求 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
 
 更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:39]:
```suggestion
在 v8.3.0 之前版本中，TiDB 仅支持在 [TiFlash MPP 模式](/tiflash/use-tiflash-mpp-mode.md)下为 `WITH ROLLUP` 语法生成有效的执行计划。因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。

从 v8.3.0 开始，TiDB 支持在不包含 TiFlash 节点的集群中为 `WITH ROLLUP` 语法生成有效的执行计划。
```

File: functions-and-operators/group-by-modifier.md [hunk 2]
```diff
@@ -51,7 +51,7 @@ CREATE TABLE bank
     profit  DECIMAL(13, 7)
 );
 
-ALTER TABLE bank SET TIFLASH REPLICA 1; -- 为该表添加一个 TiFlash 副本
+ALTER TABLE bank SET TIFLASH REPLICA 1; -- MPP 模式可以先为该表添加一个 TiFlash 副本
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:54]:
```suggestion
ALTER TABLE bank SET TIFLASH REPLICA 1; -- 在 TiFlash MPP 模式下，为该表添加一个 TiFlash 副本
```

File: functions-and-operators/group-by-modifier.md [hunk 3]
```diff
@@ -163,6 +163,23 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
 以下为一个执行计划示例：
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:163]:
```suggestion
以下为不包含 TiFlash 节点的 TiDB 集群的执行计划示例，其中 `Expand` 算子的 `task` 为 `root`，表示 `Expand` 算子在 TiDB 中执行：
```

File: functions-and-operators/group-by-modifier.md [hunk 4]
```diff
@@ -163,6 +163,23 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
...
 
 ```sql
+-- 这个是 TiDB Root 模式下的执行计划
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:166]:
```suggestion
```

File: functions-and-operators/group-by-modifier.md [hunk 5]
```diff
@@ -163,6 +163,23 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
...
+tidb> explain SELECT year, month, grouping(year), grouping(month), SUM(profit) AS profit FROM bank GROUP BY year, month WITH ROLLUP;
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:167]:
```suggestion
EXPLAIN SELECT year, month, grouping(year), grouping(month), SUM(profit) AS profit FROM bank GROUP BY year, month WITH ROLLUP;
```

File: functions-and-operators/group-by-modifier.md [hunk 6]
```diff
@@ -163,6 +163,23 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
...
+|       └─TableReader_16         | 3.00    | root      |               | data:TableFullScan_15                                                                                                                                                                                                                |
+|         └─TableFullScan_15     | 3.00    | cop[tikv] | table:bank    | keep order:false, stats:pseudo                                                                                                                                                                                                       |
++--------------------------------+---------+-----------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:180]:
```suggestion

以下为 TiFlash MPP 模式下的执行计划示例，其中 `Expand` 算子的 `task` 为 `mpp[tiflash]`，表示 `Expand` 算子在 TiFlash 中执行：

```

File: functions-and-operators/group-by-modifier.md [hunk 7]
```diff
@@ -163,6 +163,23 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
...
+```sql
+-- 这个是 TiDB MPP 模式下的执行计划
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:182]:
```suggestion
```

File: functions-and-operators/group-by-modifier.md [hunk 8]
```diff
@@ -163,6 +163,23 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
...
 explain SELECT year, month, grouping(year), grouping(month), SUM(profit) AS profit FROM bank GROUP BY year, month WITH ROLLUP;
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:183]:
```suggestion
EXPLAIN SELECT year, month, grouping(year), grouping(month), SUM(profit) AS profit FROM bank GROUP BY year, month WITH ROLLUP;
```

File: functions-and-operators/group-by-modifier.md [hunk 9]
```diff
@@ -34,7 +34,7 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
 
 ## 准备条件
 
-目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
+目前，TiDB 支持在 TiDB 自身和 TiFlash MPP 模式下为 `WITH ROLLUP` 语法生成有效的执行计划，可以在执行计划中通过观察 `EXPAND` 算子存储引擎属性来辨别。`EXPAND` 的 MPP 属性要求 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:37]: Please remove "TiDB 目前仅在 MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划" from the following note:
https://github.com/pingcap/docs-cn/blob/d5a8ee21da185208f47ba380872a5db6bcd4799b/explain-aggregation.md?plain=1#L187-L189

**Review by windtalker — APPROVED** [review]
LGTM

File: functions-and-operators/group-by-modifier.md [hunk 10]
```diff
@@ -34,9 +34,9 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
 
 ## 准备条件
 
-目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
+在 v8.3.0 之前版本中，TiDB 仅支持在 [TiFlash MPP 模式](/tiflash/use-tiflash-mpp-mode.md)下为 `WITH ROLLUP` 语法生成有效的执行计划。因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
 
-更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
+从 v8.3.0 开始，TiDB 支持在不包含 TiFlash 节点的集群中为 `WITH ROLLUP` 语法生成有效的执行计划，可以在执行计划中通过观察 `EXPAND` 算子存储引擎属性来辨别。
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:39]:
```suggestion
从 v8.3.0 开始，TiDB 支持在不包含 TiFlash 节点的集群中为 `WITH ROLLUP` 语法生成有效的执行计划。

你可以通过执行计划中 `Expand` 算子的 `task` 属性判断执行 `Expand` 算子的存储引擎。更多信息，请参考[如何阅读 ROLLUP 的执行计划](#如何阅读-rollup-的执行计划)。
```

**Review by Oreoxmt — APPROVED** [review]

File: functions-and-operators/group-by-modifier.md [hunk 11]
```diff
@@ -34,9 +34,11 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
 
 ## 准备条件
 
-目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
+在 v8.3.0 之前版本中，TiDB 仅支持在 [TiFlash MPP 模式](/tiflash/use-tiflash-mpp-mode.md)下为 `WITH ROLLUP` 语法生成有效的执行计划。因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
 
-更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
+从 v8.3.0 开始，TiDB 支持在不包含 TiFlash 节点的集群中为 `WITH ROLLUP` 语法生成有效的执行计划。
```
Comment (elsa0520) [inline functions-and-operators/group-by-modifier.md:39]:
```suggestion
从 v8.3.0 开始，TiDB 支持在所有情况下（包含 TiDB, TiFlash）集群中为 `WITH ROLLUP` 语法生成有效的执行计划。
```

File: functions-and-operators/group-by-modifier.md [hunk 12]
```diff
@@ -34,9 +34,11 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
...
+
+你可以通过执行计划中 `Expand` 算子的 `task` 属性判断执行 `Expand` 算子的存储引擎。更多信息，请参考[如何阅读 ROLLUP 的执行计划](#如何阅读-rollup-的执行计划)。
 
 ## 使用示例
```
Comment (elsa0520) [inline functions-and-operators/group-by-modifier.md:44]: 以下为在 tiflash 上使用 expand 算子的例子

File: functions-and-operators/group-by-modifier.md [hunk 13]
```diff
@@ -34,9 +34,11 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
 
 ## 准备条件
 
-目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
+在 v8.3.0 之前版本中，TiDB 仅支持在 [TiFlash MPP 模式](/tiflash/use-tiflash-mpp-mode.md)下为 `WITH ROLLUP` 语法生成有效的执行计划。因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
 
-更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
+从 v8.3.0 开始，TiDB 支持在所有情况下（包含 TiDB, TiFlash）集群中为 `WITH ROLLUP` 语法生成有效的执行计划。
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:39]:
```suggestion
从 v8.3.0 开始，上述限制已被移除。无论 TiDB 集群是否包含 TiFlash 节点，TiDB 都支持为 `WITH ROLLUP` 语法生成有效的执行计划。
```

File: functions-and-operators/group-by-modifier.md [hunk 14]
```diff
@@ -160,10 +162,27 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
```
Comment (qiancai) [inline functions-and-operators/group-by-modifier.md:162]: 因为 v8.3 起 Expand 算子也可以在 TiDB 中执行，L161 这里，要不要修改为这样：
```
多维度数据聚合使用了 `Expand` 算子来复制数据以满足多维度分组的需求，每个复制的数据副本都对应一个特定维度的分组。`Expand` 算子能够快速地在多个节点之间重新组织和计算大量的数据，充分利用每个节点的计算能力。
```

File: functions-and-operators/group-by-modifier.md [hunk 15]
```diff
@@ -34,9 +34,11 @@ SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
 
 ## 准备条件
 
-目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。
+在 v8.3.0 之前版本中，TiDB 仅支持在 [TiFlash MPP 模式](/tiflash/use-tiflash-mpp-mode.md)下为 `WITH ROLLUP` 语法生成有效的执行计划。因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
 
-更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。
+从 v8.3.0 开始，上述限制已被移除。无论 TiDB 集群是否包含 TiFlash 节点，TiDB 都支持为 `WITH ROLLUP` 语法生成有效的执行计划。
+
+你可以通过执行计划中 `Expand` 算子的 `task` 属性判断执行 `Expand` 算子的存储引擎。更多信息，请参考[如何阅读 ROLLUP 的执行计划](#如何阅读-rollup-的执行计划)。
```
Comment (qiancai) [inline functions-and-operators/group-by-modifier.md:41]:
```suggestion
你可以通过执行计划中 `Expand` 算子的 `task` 属性判断执行 `Expand` 算子的是 TiDB 还是 TiFlash。更多信息，请参考[如何阅读 ROLLUP 的执行计划](#如何阅读-rollup-的执行计划)。
```

Comment (qiancai) [reply functions-and-operators/group-by-modifier.md:41]: reason: TiDB 可能不算存储引擎？

**Review by qiancai — APPROVED** [review]
Rest LGTM

File: functions-and-operators/group-by-modifier.md (hunk 14 above)
Comment (AilinKid) [reply functions-and-operators/group-by-modifier.md:162]: 我改成这样了： ”多维度数据聚合使用了 `Expand` 算子来复制数据以满足多维度分组的需求，每个复制的数据副本都对应一个特定维度的分组。特别的在 MPP 模式下，`Expand` 算子能够利于数据 shuffle 快速地在多个节点之间重新组织和计算大量的数据，充分利用每个节点的计算能力；而在 TiDB 模式下，`Expand` 算子因为只在 TiDB 单节点上执行，数据冗余会根据 `Grouping Sets` 的数量而膨胀。“ PTAL

File: functions-and-operators/group-by-modifier.md [hunk 16]
```diff
@@ -156,14 +158,31 @@ SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(
 
 ## 如何阅读 ROLLUP 的执行计划
 
-多维度数据聚合使用了 `Expand` 算子来复制数据以满足多维度分组的需求，每个复制的数据副本都对应一个特定维度的分组。通过 MPP 的数据 shuffle 能力，`Expand` 算子能够快速地在多个 TiFlash 节点之间重新组织和计算大量的数据，充分利用每个节点的计算能力。
+多维度数据聚合使用了 `Expand` 算子来复制数据以满足多维度分组的需求，每个复制的数据副本都对应一个特定维度的分组。特别的在 MPP 模式下，`Expand` 算子能够利于数据 shuffle 快速地在多个节点之间重新组织和计算大量的数据，充分利用每个节点的计算能力；而在 TiDB 模式下，`Expand` 算子因为只在 TiDB 单节点上执行，数据冗余会根据 `Grouping Sets` 的数量而膨胀。
```
Comment (Oreoxmt) [inline functions-and-operators/group-by-modifier.md:161]:
```suggestion
多维度数据聚合使用了 `Expand` 算子来复制数据以满足多维度分组的需求，每个复制的数据副本都对应一个特定维度的分组。在 MPP 模式下，`Expand` 算子能够利用数据 shuffle 快速地在多个节点之间重新组织和计算大量的数据，充分利用每个节点的计算能力。在不包含 TiFlash 节点的 TiDB 集群中，`Expand` 算子因为只在 TiDB 单节点上执行，数据冗余会随着维度分组 (`grouping set`) 数量的增加而增加。
```

**Pull request merged**